implementations/rust/ockam/ockam_command/src/base_command.rs

```diff
@@ -524,86 +525,139 @@ impl Display for ReadStatus {
+            if let Err(e) = RustylineHandler::run_repl(next_r, lines_s, history_file_path) {
+                eprintln!("{:?}", e);
+            }
+        });
+        RustylineHandle {
+            lines: lines_r,
+            next: next_s,
+        }
+    }
 
-                match read_result {
-                    Ok(0) => break, // EOF
-                    Ok(_n) => {
-                        if read_buffer.trim().is_empty() {
-                            continue;
-                        }
-                        let _ = tx.send(read_buffer.clone());
-                        if is_quit_sequence(&read_buffer) {
-                            break;
-                        }
-                    }
-                    Err(e) => {
-                        eprintln!("Error reading from stdin: {}", e);
+    fn run_repl(
+        mut next_r: Receiver<()>,
+        lines_s: Sender<String>,
+        history_file_path: Option<PathBuf>,
+    ) -> Result<()> {
+        let history_file_pat = Self::get_history_file_path(history_file_path)?;
+        let mut rl = Editor::new()
+            .into_diagnostic()
+            .wrap_err("Failed to initialize REPL")?;
+        rl.set_max_history_size(50).into_diagnostic()?;
+        let helper = MultiLineValidatorHelper {};
+        rl.set_helper(Some(helper));
+        rl.load_history(&history_file_pat).into_diagnostic()?;
+        loop {
+            // If we can't read from the channel, exit
+            if next_r.blocking_recv().is_none() {
+                break;
+            };
+            let readline = rl.readline("> ");
+            match readline {
+                Ok(line) => {
+                    if is_quit_sequence(&line) {
                         break;
                     }
+                    rl.add_history_entry(line.as_str()).into_diagnostic()?;
+                    let _ = lines_s.blocking_send(line);
+                    let _ = rl.save_history(&history_file_pat);
+                }
+                Err(ReadlineError::Interrupted) => break,
+                Err(ReadlineError::Eof) => break,
+                Err(err) => {
+                    eprintln!("Error reading user input: {:?}", err);
+                    break;
                 }
             }
+        }
+        Ok(())
+    }
 
-            drop(tx);
-        });
-
-        StdinHandle { tx: returned_tx }
+    fn get_history_file_path(history_file_path: Option<PathBuf>) -> Result<PathBuf> {
+        let create_history_filecreate_history_file = |path: &PathBuf| -> Result<()> {
+            if !path.exists() {
+                if let Some(parent) = path.parent() {
+                    if !parent.exists() {
+                        std::fs::create_dir_all(parent)
+                            .into_diagnostic()
+                            .wrap_err("Failed to create history directory for REPL")?;
+                    }
+                }
+                std::fs::File::create(path)
```

Comment:
Lets restrict permissions on that file to `600`

Reply:
it's done by the library already (if it has a different permission, it set it to 600)